File: ml/train/sample.py
# SPDX-License-Identifier: GPL-2.0-or-later
"""Reference sampler — numpy only, no torch. Byte-for-byte the same RNG and
top-k / softmax procedure as `ml/thos-lm/src/sample.rs`, so this and the Rust
`generate` example produce the identical token sequence from the same seed.

    python ml/train/sample.py --weights spike-1m.tlm --prompt "The " \
        --max-tokens 200 --temp 0.8 --top-k 40 --seed 1
"""

from __future__ import annotations

import argparse

import numpy as np

from tlm import numpy_forward, read_tlm

MASK64 = (1 << 64) - 1
MUL = 0x2545F4914F6CDD1D


class Rng:
    __slots__ = ("s",)

    def __init__(self, seed: int):
        self.s = seed if seed != 0 else 0x9E3779B97F4A7C15

    def next_u64(self) -> int:
        x = self.s
        x ^= x >> 12
        x ^= (x << 25) & MASK64
        x ^= x >> 27
        self.s = x
        return (x * MUL) & MASK64

    def next_f32(self) -> float:
        return (self.next_u64() >> 40) / float(1 << 24)


def pick(logits: np.ndarray, temp: float, top_k: int, rng: Rng) -> int:
    if temp <= 0.0:
        return int(np.argmax(logits))
    logits = (logits / np.float32(temp)).astype(np.float32)
    k = max(1, min(top_k, logits.size))

    order = []
    used = np.zeros(logits.size, dtype=bool)
    for _ in range(k):
        masked = np.where(used, -np.inf, logits)
        bi = int(np.argmax(masked))  # first-max on ties
        used[bi] = True
        order.append(bi)

    vals = np.array([logits[i] for i in order], dtype=np.float32)
    vals = vals - vals.max()
    e = np.exp(vals)
    probs = e / e.sum()

    u = rng.next_f32()
    acc = 0.0
    for j, p in enumerate(probs):
        acc += float(p)
        if u < acc:
            return order[j]
    return order[-1]


def main() -> None:
    ap = argparse.ArgumentParser()
    ap.add_argument("--weights", required=True)
    ap.add_argument("--prompt", default="The ")
    ap.add_argument("--max-tokens", type=int, default=200)
    ap.add_argument("--temp", type=float, default=0.8)
    ap.add_argument("--top-k", type=int, default=40)
    ap.add_argument("--seed", type=int, default=1)
    ap.add_argument("--dump-tokens", action="store_true", help="print the token id list too")
    args = ap.parse_args()

    cfg, w = read_tlm(args.weights)
    toks = list(args.prompt.encode("utf-8"))
    rng = Rng(args.seed)
    for _ in range(args.max_tokens):
        ctx = toks[-cfg.block_size:]
        logits = numpy_forward(cfg, w, ctx)
        toks.append(pick(logits, args.temp, args.top_k, rng))

    text = bytes(t for t in toks if t < 256).decode("utf-8", errors="replace")
    print(text)
    if args.dump_tokens:
        print("\n# tokens:", " ".join(map(str, toks)))


if __name__ == "__main__":
    main()

File: ml/thos-lm/src/sample.rs
// SPDX-License-Identifier: GPL-2.0-or-later
//! Deterministic token sampling. The RNG (xorshift64*) and the top-k / softmax
//! procedure are byte-for-byte the same as `ml/train/sample.py`, so a Rust run
//! and the Python reference produce the identical token sequence from the same
//! seed.

use alloc::vec::Vec;

use crate::math::{argmax, softmax};
use crate::model::Model;

/// xorshift64* — tiny, deterministic, good enough for sampling.
pub struct Rng {
    state: u64,
}

impl Rng {
    pub fn new(seed: u64) -> Self {
        Rng { state: if seed == 0 { 0x9E37_79B9_7F4A_7C15 } else { seed } }
    }
    pub fn next_u64(&mut self) -> u64 {
        let mut x = self.state;
        x ^= x >> 12;
        x ^= x << 25;
        x ^= x >> 27;
        self.state = x;
        x.wrapping_mul(0x2545_F491_4F6C_DD1D)
    }
    /// Uniform in `[0, 1)` from the top 24 bits.
    pub fn next_f32(&mut self) -> f32 {
        ((self.next_u64() >> 40) as f32) / ((1u32 << 24) as f32)
    }
}

#[derive(Debug, Clone, Copy)]
pub struct SamplerConfig {
    pub temperature: f32,
    pub top_k: usize,
    pub max_tokens: usize,
}

impl Default for SamplerConfig {
    fn default() -> Self {
        SamplerConfig { temperature: 0.8, top_k: 40, max_tokens: 64 }
    }
}

pub struct Sampler {
    cfg: SamplerConfig,
    rng: Rng,
}

impl Sampler {
    pub fn new(cfg: SamplerConfig, seed: u64) -> Self {
        Sampler { cfg, rng: Rng::new(seed) }
    }

    /// One next token from `logits` (consumed/mutated as scratch).
    pub fn pick(&mut self, logits: &mut [f32]) -> u16 {
        if self.cfg.temperature <= 0.0 {
            return argmax(logits) as u16;
        }
        let inv_t = 1.0 / self.cfg.temperature;
        for x in logits.iter_mut() {
            *x *= inv_t;
        }
        let k = self.cfg.top_k.clamp(1, logits.len());

        // Deterministic top-k: pick the max k times, ties -> lowest index.
        let mut chosen: Vec<(usize, f32)> = Vec::with_capacity(k);
        let mut used = alloc::vec![false; logits.len()];
        for _ in 0..k {
            let mut bi = usize::MAX;
            let mut bv = f32::NEG_INFINITY;
            for (i, &v) in logits.iter().enumerate() {
                if !used[i] && v > bv {
                    bv = v;
                    bi = i;
                }
            }
            used[bi] = true;
            chosen.push((bi, bv));
        }

        let mut probs: Vec<f32> = chosen.iter().map(|&(_, v)| v).collect();
        softmax(&mut probs);

        let u = self.rng.next_f32();
        let mut acc = 0.0f32;
        for (j, &p) in probs.iter().enumerate() {
            acc += p;
            if u < acc {
                return chosen[j].0 as u16;
            }
        }
        chosen[chosen.len() - 1].0 as u16
    }

    /// Autoregressively extend `tokens` in place by up to `max_tokens`, clamping
    /// the context to the model's `block_size`.
    pub fn generate(&mut self, model: &Model, tokens: &mut Vec<u16>) {
        let bs = model.cfg.block_size;
        for _ in 0..self.cfg.max_tokens {
            let start = tokens.len().saturating_sub(bs);
            let mut logits = model.forward(&tokens[start..]);
            let next = self.pick(&mut logits);
            tokens.push(next);
        }
    }
}
